### DISCUSSION TITLE
00_pytorch_fundamentals_exercises Qn 9

### DISCUSSION THREAD

**Discussion by chewyh26** [mrdbourke/pytorch-deep-learning#181] [Q&A] [answered]
Hello, I have 1 question regarding Qn 9, Find the maximum and minimum index values of the output of 7.

The output of qn7 is a 2-dimensional tensor

```
(tensor([[0.3647, 0.4709],
         [0.5184, 0.5617]], device='cuda:0'), torch.Size([2, 2]))
```
but the output index values are not representative of the position of the maximum and/or minimum value.

For example, in the case of maximum value, the index position should be [1, 1] or [1][1] to arrive at the maximum value, but it shows the output as 3, i.e. (tensor(3, device='cuda:0').

As the lesson video only shows 1-dimensional tensor as demonstration, I am not sure how the code can be modified to get the accurate index position. I refer to the solution and the output is also shown as 3 which isn't correct logically.

Code:

```python
# Find arg max
tensor_C.argmax()

# Find arg min
tensor_C.argmin()
tensor_C.argmax(), tensor_C.argmin()
```

**Comment by ghost** [comment]
### You should indicate the **dim** from which you want to find the **min** or **max**

![image](https://user-images.githubusercontent.com/87841705/202413201-7a74add6-a402-49ed-995b-6af4fec0a55f.png)

### First time, it may be a bit confusing. I hope this explanation will help you:

![image](https://user-images.githubusercontent.com/87841705/202413674-3c815c7a-7a2f-4619-86c7-f0111d05ee9e.png)

### Moreover, it is more understandable if you use bigger matrices

![image](https://user-images.githubusercontent.com/87841705/202414137-9a872d92-59ed-49ed-9862-d033cb9a5e33.png)

![image](https://user-images.githubusercontent.com/87841705/202415317-22b2462c-531d-42ca-8b29-caeefc17f142.png)

**Reply by chewyh26** [reply]
thank you so much!

**Reply by ghost** [reply]
Welcome )

**Reply by chewyh26** [reply]
for the case of 2 dimensional tensor you had shown, I presume it compares only among the different columns if dim set to be 0 and different rows if i set dim = 1. Eg. for the higher dimension tensor, say torch.size [(4,4)], using arg.max, u had 4 different highest values in each columns, wherein the screenshots seem to tell me that it is comparing among the same columns if dim set to 1, and giving only the 4 respective highest element indexes.

what if I want to compare the highest among all elements in the 4X4 dimensional tensor, giving me just 1 index position?  I understand that .argmax() can give an output of the single highest index value, but I have to lay the input 1 by 1 in 1 dimension and counting it to the correct position. eg. 
code:
tensor_Z =torch.rand(4,4)
tensor_Z.argmax(), tensor_Z

output:

(tensor(11), tensor([[0.7447, 0.7213, 0.4414, 0.5550],
         [0.6361, 0.1081, 0.3305, 0.5196],
         [0.2147, 0.2816, 0.6679, 0.7878],
         [0.5070, 0.3055, 0.7738, 0.2259]]))

it gives 11th position but i have to count from 0: 0.7447, 1: 0.7213,. 4: 0.6361... 11th: 0.7878 which is the highest..

**Comment by chewyh26** [comment]
I wrote a random example to summarise and clarify my last reply
<img width="392" alt="code" src="https://user-images.githubusercontent.com/66502176/202431084-93e7a6de-757d-4f52-ae2c-008d5f4c3d33.png">

**Reply by ghost** [reply]
I don't know whether PyTorch has a such functionality. But I wrote small function to find position

![image](https://user-images.githubusercontent.com/87841705/202454850-535e07fa-9bd0-43b9-b70c-5a6f89c861d7.png)

![image](https://user-images.githubusercontent.com/87841705/202454898-e50a234d-e6d5-451f-8221-bdad6b06da17.png)

I have not finished course and I don't know whether the function like `stride()` is explained. Because, I learnt it from other source

**Reply by ghost** [reply]
Moreover, my function will not work in many conditions. But in general, more or less works )))

**Reply by chewyh26** [reply]
> Moreover, my function will not work in many conditions. But in general, more or less works )))

Hello, thanks for answering my question, I am not familiar with the function stride(), like what it does and can't comprehend the small function you'd written. Referring to the pytorch documentation and some other discussions, I do understand abit that it will output tuple and the jump steps necessary to get to other row depending on the dimensions of tensor... enumerating list of tuple(s) --> each stride output is different and correspond to the different dimensions of tensors since dim is not specified? And not sure how line 5- line 6 of the code works. will you perhaps be able to dissect the code to inform the rationale and also share the source you had learnt from? Thanks alot for your help!

**Comment by ghost** [comment]
I read it from the book "Deep Learning with PyTorch". `stride()` and `storage()` are pretty confusing topics and out of "fundamentals of PyTorch". But still I can explain as much as I can if you want. But I think it is not important right now

**Reply by chewyh26** [reply]
thank you for sharing the resource!

**Comment by mrdbourke** [comment, marked as answer]
Hi @chewyh26,

First, fantastic question, the exercise + answers are a little confusing now I read them again.

And @JahoNematov thank you for your help and offerings.

I've got a hacked-together solution from some research online.

Namely via the use of NumPy's [`unravel_index()`](https://numpy.org/doc/stable/reference/generated/numpy.unravel_index.html). 

Again this is where the confusion comes in because this is a PyTorch course not a NumPy course. But for *most* NumPy functions there is a PyTorch equivalent.

You can achieve the desired outcome of question 9 "Find the maximum and minimum index values of the output of 7" via the following:

```python
import torch
x = torch.tensor([[0.3647, 0.4709],
                  [0.5184, 0.5617]])
x
```
```
>>> 
tensor([[0.3647, 0.4709], 
        [0.5184, 0.5617]])
```
Now to find the max indexes:

```python
def find_max_index_with_numpy(x):
  # Find the max value with torch.max()
  max = torch.max(x)
  # Find the max indices with unravel_index()
  max_index = np.unravel_index(x.argmax(), x.shape)
  # Index into the target tensor with the max indices
  max_with_indexing = x[max_index]

  print(f"Max with torch.max(): {max}")
  print(f"Max index: {max_index}")
  print(f"Max with indexing: {max_with_indexing}")
  print(f"Do they match? {max == max_with_indexing}")

find_max_index_with_numpy(x)
```
```
>>> 
Max with torch.max(): 0.5616999864578247
Max index: (1, 1)
Max with indexing: 0.5616999864578247
Do they match? True
```

Teaching the insides of `numpy.unravel_index()` is beyond the scope of the course but I'm glad you bought this issue up.

These resources helped me to learn about this:M
* [What is an intuitive explanation of np.unravel_index?](https://stackoverflow.com/questions/48135736/what-is-an-intuitive-explanation-of-np-unravel-index)
* [Argmax of numpy array returning non-flat indices](https://stackoverflow.com/questions/9482550/argmax-of-numpy-array-returning-non-flat-indices), and more specifically this answer: https://stackoverflow.com/a/9483964/7900723

Let me know if you have anymore questions :)

**Reply by chewyh26** [reply]
Thank you!!
